libsolidity/analysis/TypeChecker.cpp

```diff
@@ -4017,24 +4052,72 @@ void TypeChecker::endVisit(Literal const& _literal)
 						_literal.location(),
 						"Functions that take 2 arguments can only be used as literal suffixes for rational numbers."
 					);
-				else
+				else if (
+					!!dynamic_cast<IntegerType const*>(suffixFunctionType->parameterTypes().at(0)) &&
+					!!dynamic_cast<IntegerType const*>(suffixFunctionType->parameterTypes().at(1))
+				)
 				{
 					auto&& [mantissa, exponent] = literalRationalType->mantissaExponent();
 					// This was already validated in visit(Literal) but the error is not fatal.
 					if (!mantissa || !exponent)
 						solAssert(!m_errorReporter.errors().empty());
-					else if (
-						!mantissa->isImplicitlyConvertibleTo(*suffixFunctionType->parameterTypes().at(0)) ||
-						!exponent->isImplicitlyConvertibleTo(*suffixFunctionType->parameterTypes().at(1))
-					)
-						// TODO: Is this triggered when the argument is out of range? Test.
-						parameterTypeMessage = "The type of the literal cannot be converted to the parameters of the suffix function.";
+					// TODO: Is this triggered when the argument is out of range? Test.
+					else
+					{
+						//optional<string> mantissaOrExponentErrorMessage;
+						vector<string> mantissaOrExponentErrorMessages;
+						if (!mantissa->isImplicitlyConvertibleTo(*suffixFunctionType->parameterTypes().at(0)))
+							mantissaOrExponentErrorMessages.emplace_back(
+								fmt::format(
+									"The mantissa is out of range of type {}",
+									suffixFunctionType->parameterTypes().at(0)->toString(true)
+								)
+							);
+						if (!exponent->isImplicitlyConvertibleTo(*suffixFunctionType->parameterTypes().at(1)))
+						{
+							mantissaOrExponentErrorMessages.emplace_back(
+								fmt::format(
+									"The exponent is out of range of type {}",
+									suffixFunctionType->parameterTypes().at(1)->toString(true)
+								)
+							);
+						}
+
+						if (!mantissaOrExponentErrorMessages.empty())
+							parameterTypeMessage =
+								"This number cannot be decomposed into a mantissa and decimal exponent "
+								"that fit the range of parameters of the suffix function. " +
+								joinHumanReadable(mantissaOrExponentErrorMessages, ".") + ".";
+					}
+				}
+				else
+				{
+					// visit(FunctionDefinition) should have spotted if any of the parameters is not Integer
+					solAssert(!m_errorReporter.errors().empty());
+					parameterTypeMessage = "This number is suffixed by a function with wrong type of parameters.";
```

Comment:
I think we can drop this one. The assert is enough. Otherwise the message should be reworded to say that the literal cannot be converted to the parameters (the fact that function has wrong types was already pointed out by another message). But looking at the test cases, it does not seem to add much.